sycl/source/detail/plugin.hpp

```diff
@@ -184,11 +184,67 @@ class plugin {
   void *getLibraryHandle() { return MLibraryHandle; }
   int unload() { return RT::unloadPlugin(MLibraryHandle); }
 
+  // return the index of PiPlatforms.
+  // If not found, add it and return its index.
+  int getPlatformId(RT::PiPlatform Platform) {
+    if (PiPlatforms) {
+      auto It = std::find(PiPlatforms->begin(), PiPlatforms->end(), Platform);
+      if (It != PiPlatforms->end())
+        return It - PiPlatforms->begin();
+
+      PiPlatforms->push_back(Platform);
+      LastDeviceIds.push_back(0);
+      return PiPlatforms->size() - 1;
+    }
+    return -1;
+  }
+  // Device ids are consecutive across platforms within a plugin.
+  // We need to return the same starting index for the given platform.
+  // So, instead of returing the last device id of the given platform,
+  // return the last device id of the predecessor platform.
+  int getStartingDeviceId(RT::PiPlatform Platform) {
+    if (!PiPlatforms)
+      PiPlatforms = new std::vector<RT::PiPlatform>;
+    int PlatformId = getPlatformId(Platform);
+    if (PlatformId <= 0)
+      return 0;
+
+    return LastDeviceIds[PlatformId - 1];
+  }
+  // set the id of the last device for the given platform
+  void setLastDeviceId(RT::PiPlatform Platform, int Id) {
+    int PlatformId = getPlatformId(Platform);
+    LastDeviceIds[PlatformId] = Id;
+  }
+  // reset all last device ids to zeros
+  void resetLastDeviceIds() {
+    std::fill(LastDeviceIds.begin(), LastDeviceIds.end(), 0);
+  }
+
+  void resetPiPlatforms() {
+    delete PiPlatforms;
+    PiPlatforms = nullptr;
+  }
```

Comment:
Still do not understand why this function is needed. Please add a comment clarifying that.
Related to another comment.

Reply:
To avoid PiPlatforms to be destroyed twice, we need to reset to a nullptr.
PiPlatforms are owned by plugins and they will be destroyed by the plugin.